lxc/completion.go

```diff
@@ -250,44 +245,144 @@ func (g *cmdGlobal) cmpInstanceKeys(instanceName string) ([]string, cobra.ShellC
 		return nil, cobra.ShellCompDirectiveError
 	}
 
+	// Complete keys based on instance type.
 	instanceType := instance.Type
 
-	if instanceType == "container" {
-		for k := range instancetype.InstanceConfigKeysContainer {
-			keys = append(keys, k)
-		}
-	} else if instanceType == "virtual-machine" {
-		for k := range instancetype.InstanceConfigKeysVM {
-			keys = append(keys, k)
-		}
+	metadataConfiguration, err := client.GetMetadataConfiguration()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	instanceConfig, ok := metadataConfiguration.Configs["instance"]
+	if !ok {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysAny {
-		keys = append(keys, k)
+	// Pre-allocate configKeys slice capacity.
+	keyCount := 0
+	for _, field := range instanceConfig {
+		keyCount += len(field.Keys)
 	}
 
-	return keys, cmpDirectives
+	configKeys := make([]string, 0, keyCount)
+
+	for _, field := range instanceConfig {
+		for _, key := range field.Keys {
+			for configKey, configKeyField := range key {
+				if strings.HasSuffix(configKey, "*") {
+					configKey = strings.Trim(configKey, "*")
+				}
+
+				if instanceType == string(api.InstanceTypeContainer) && configKeyField.Condition == "container" {
+					configKeys = append(configKeys, configKey)
+				} else if instanceType == string(api.InstanceTypeVM) && configKeyField.Condition == "virtual machine" {
+					configKeys = append(configKeys, configKey)
+				}
+
+				// InstanceTypeAny config keys.
+				if configKeyField.Condition == "" {
+					configKeys = append(configKeys, configKey)
+				}
```

Comment:
Doing this `if` check before the instance type would be a tad faster.

Reply:
Great suggestion, thanks!